std/utf.d

```diff
@@ -2420,6 +2469,167 @@ unittest // bugzilla 12923
     }());
 }
 
+/++
+    Returns whether the given range of characters is well-formed Unicode or not.
+
+    Params:
+        range = The range of charaters to validate.
+        index = The index (in code units) where the failed character was (or `0`
+                if the result is `true`).
+  +/
+pragma(inline, true) bool isValidUnicode(R)(R range)
+    if (isInputRange!R && isSomeChar!(ElementType!R) && !isInfinite!R)
+{
+    size_t failedIndex;
+    return _isValidUnicodeImplTrustWrap!false(range, failedIndex);
+}
+
+/// Ditto
+pragma(inline, true) bool isValidUnicode(R)(R range, out size_t failedIndex)
+    if (isInputRange!R && isSomeChar!(ElementType!R) && !isInfinite!R)
+{
+    return _isValidUnicodeImplTrustWrap!true(range, failedIndex);
+}
+
+pragma(inline, true)
+private bool _isValidUnicodeImplTrustWrap(bool getIndex, R)(R range, out size_t failedIndex) @trusted
+    if(isSomeString!R)
+{
+    return _isValidUnicodeImpl!getIndex(range, failedIndex);
+}
+
+pragma(inline, true)
+private bool _isValidUnicodeImplTrustWrap(bool getIndex, R)(R range, out size_t failedIndex)
+    if(!isSomeString!R)
+{
+    return _isValidUnicodeImpl!getIndex(range, failedIndex);
+}
+
+private bool _isValidUnicodeImpl(bool getIndex, R)(R range, out size_t failedIndex)
+    if(isSomeString!R || isRandomAccessRange!R)
+{
+    immutable len = range.length;
+    for (size_t i = 0; i < len;)
+    {
+        if (range[i] < codeUnitLimit!R)
+            ++i;
+        else if(!decodeImpl!(true, DecodeType.onlyValidate)(range, i))
+        {
+            failedIndex = i;
+            return false;
+        }
+    }
+    return true;
+}
+
+private bool _isValidUnicodeImpl(bool getIndex, R)(R range, out size_t failedIndex)
+    if(!isSomeString!R && !isRandomAccessRange!R)
+{
+    for (size_t i = 0; !range.empty; ++i)
+    {
+        immutable c = range.front;
+        if(c < codeUnitLimit!R)
+            range.popFront();
+        else if(!decodeImpl!(false, DecodeType.onlyValidate)(range, i))
```

Comment:
I think you're unintentionally incrementing `i` twice in one iteration whenever `decodeImpl()` passes, which will throw off the `failedIndex` reporting.

To test this, make a `dchar` string with a few code points from the astral planes, followed by an invalid one, and see if the resulting `failedIndex` is within the bounds of the string or not. (I'd like to see something like this in a `unittest` block below, even if this function already works OK.)